File: java/src/main/java/benchmark/Fidelity.java
package benchmark;

import java.lang.reflect.Array;
import java.util.List;
import java.util.Objects;

/** Semantic equality for suite fixtures (float tolerance). */
public final class Fidelity {
  private Fidelity() {}

  public static boolean check(Object expected, Object actual) {
    if (expected == actual) return true;
    if (expected == null || actual == null) return false;
    if (expected instanceof List<?> el && actual instanceof List<?> al) {
      if (el.size() != al.size()) return false;
      for (int i = 0; i < el.size(); i++) {
        if (!check(el.get(i), al.get(i))) return false;
      }
      return true;
    }
    if (expected.getClass().isArray() && actual.getClass().isArray()) {
      int n = Array.getLength(expected);
      if (n != Array.getLength(actual)) return false;
      for (int i = 0; i < n; i++) {
        if (!check(Array.get(expected, i), Array.get(actual, i))) return false;
      }
      return true;
    }
    if (expected instanceof Double || expected instanceof Float
        || actual instanceof Double || actual instanceof Float) {
      if (expected instanceof Number en && actual instanceof Number an) {
        return Math.abs(en.doubleValue() - an.doubleValue()) <= 1e-9;
      }
    }
    // Prefer equals when both sides implement meaningful equals (our POJOs do).
    if (expected.equals(actual)) return true;
    // Jackson/Gson may produce LinkedHashMap for untyped decode — not used if typed.
    return Objects.deepEquals(expected, actual);
  }
}

File: kotlin/src/main/kotlin/benchmark/Fidelity.kt
package benchmark

import java.lang.reflect.Array as JArray

/** Semantic equality for suite fixtures (float tolerance). */
object Fidelity {
    fun check(expected: Any?, actual: Any?): Boolean {
        if (expected === actual) return true
        if (expected == null || actual == null) return false
        if (expected is List<*> && actual is List<*>) {
            if (expected.size != actual.size) return false
            for (i in expected.indices) {
                if (!check(expected[i], actual[i])) return false
            }
            return true
        }
        if (expected.javaClass.isArray && actual.javaClass.isArray) {
            val n = JArray.getLength(expected)
            if (n != JArray.getLength(actual)) return false
            for (i in 0 until n) {
                if (!check(JArray.get(expected, i), JArray.get(actual, i))) return false
            }
            return true
        }
        if (expected is Double || expected is Float || actual is Double || actual is Float) {
            if (expected is Number && actual is Number) {
                return kotlin.math.abs(expected.toDouble() - actual.toDouble()) <= 1e-9
            }
        }
        if (expected == actual) return true
        return expected == actual
    }
}
